rfd/0135-kube-apps-discovery.md

```diff
@@ -0,0 +1,199 @@
+
+## Details
+
+Kubernetes app discovery will build upon existing infrastructure for other discovery capabilities that are already there. We will add another
+type of matchers to the discovery service configuration - `kubernetes`. When discovery service sees this, it will periodically (once every 5 minutes)
+poll Kubernetes cluster, inside which it is running, for a list of services. Based on this list, it will form a list of Teleport apps resources
+and update it on the backend, constantly keeping the dynamic apps state up to date. Writing to the backend will be handled
+by already existing mechanism in our code - reconciler. Reconciler will compare given fresh list of resources and resources we currently have
+on the backend, and then will decide on appropriate actions (create/update/delete) that are required on
+the backend to change resources to the desired state.
+The Teleport app service that runs in the same Kubernetes cluster will then react to those changes, and will start/stop proxying
+apps that are selected by appropriate labels.
+
+Name of the created Teleport app will consist of Kubernetes service name, namespace and cluster name, like this:
+`$SERVICE_NAME-$CLUSTER_NAME-$NAMESPACE`. This is done to avoid naming collisions in discovered apps. Though app names
+will be long in that case, there's ongoing work to improve UX when working with long resources names in Teleport, see [RFD 129](https://github.com/gravitational/teleport/pull/27258).
+
+To allow for finer control over exposed service transformation into Teleport app Kubernetes annotations will be used.
+By default we will be exposing services as `tcp` apps, since it's the most general type, but annotation or exposed port's `appProtocol` field
+can be used to specify which protocol to use for the app's URI. 
+
+If Kubernetes service has multiple ports in use, we will expose each port as a separate app, with name of the app including the port's name
+specified in the service. For example, if a service has port 443 exposed with name `tls` and also port `5432` with name `postgres`
+we will create two apps: `testApp-tls-main-cluster-prod`, `testApp-postgres-main-cluster-prod`. Names for ports are mandatory in Kubernetes if more
+than one port is exposed for the service, so we always will be able to use it if we need it. User will be able to use annotation to specify
+preferred port, in that case only app with this port will be created, without port info in the app name.
+
+In order for discovery service to be able to dynamically create/update/delete apps we will add Read/Write permissions 
+for the apps to the built-in system role `Discovery`.
+
+Labels from the Kubernetes service will be copied to the corresponding Teleport app.
+
+## UX
+
+### Discovery
+
+Discovery service will use new `kubernetes` matchers to periodically poll list of services inside the Kubernetes cluster.
```

Comment:
What happens if a service contains no endpoint? Do we still discover it? Issue a warning? e.g. no pods are ready, the underlying deployment has been scaled down, and the service matches no pods because of a label selector issue.

Reply:
IIRC app service performs health checks so it will ensure the service is reachable and ?healthy?  before heart beating the `app_server`